### PR TITLE
Use os.ReadDir instead of filepath.Walk in netns cleanup

### PR DESCRIPTION
**Pull request by thaJeztah**
- Use `ioutil.ReadDir()` instead of `filepath.Walk()`, because
`filepath.Walk()` does a lot of extra things we don't need
- Log errors on directory removal (omitting `os.IsNotExist`)
- Touch-up error messages for consistency

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @kolyshkin @cpuguy83 @fcrisciani PTAL; not entirely sure if this is correct, but opened this to discuss 👍

File: integration-cli/daemon/daemon_unix.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ func cleanupExecRoot(c *check.C, execRoot string) {
 	// new exec root.
 	netnsPath := filepath.Join(execRoot, "netns")
 	filepath.Walk(netnsPath, func(path string, info os.FileInfo, err error) error {
-		if err := unix.Unmount(path, unix.MNT_FORCE); err != nil {
+		if err := unix.Unmount(path, unix.MNT_FORCE); err != nil && err != unix.EINVAL {
```
Comment (thaJeztah) [inline integration-cli/daemon/daemon_unix.go]: Should we also ignore `ENOENT`? 

```
04:02:02 unmount of /tmp/docker-execroot/d1825ad513bf4/netns failed: invalid argument
04:02:02 unmount of /tmp/docker-execroot/d1825ad513bf4/netns failed: no such file or directory
```

**Comment by thaJeztah** [comment]
hm, this one is flaky /cc @selansen

```
13:44:44 --- FAIL: TestServiceWithPredefinedNetwork (11.75s)
13:44:44 	daemon.go:283: [dce2b71fc535a] waiting for daemon to start
13:44:44 	daemon.go:315: [dce2b71fc535a] daemon started
13:44:44 	service_test.go:52: timeout hit after 10s: task count at 1 waiting for 0
13:44:44 	daemon.go:273: [dce2b71fc535a] exiting daemon
```

**Comment by selansen** [comment]
It didnt fail for me and number of instance can never be 0. will check it
again.

**Comment by selansen** [comment]
Interesting . I didnt see this issue while I test and  when CI ran against
my PR.
Let me see if I can reproduce it locally. Else will  check for service
instead of task and see  if it helps.

File: integration-cli/daemon/daemon_unix.go (hunk 1 above)
Comment (kolyshkin) [reply integration-cli/daemon/daemon_unix.go]: Yes, I think ENOENT can be returned due to races and should be ignored.

Oh, and `filepath.Walk()` should not be used here as it's recursive and does lots of extra things we don't need. Using `os.Readdirnames()` should be sufficient (similar to https://github.com/moby/moby/pull/36307).

**Comment by kolyshkin** [comment]
The `MNT_FORCE` flag comes from commit 99a98ccc14 (PR #25962). The only other use of MNT_FORCE is in libnetwork (https://github.com/docker/libnetwork/commit/8a6b507763478b87da2d33776af85b9732b90272#diff-50ec864109e0293aa03f973d35485e1dR396), both are by @mrjana (Jana, maybe you can shed some light on why this option is used?)

FS type is `nsfs` in both cases. My limited testing shows the flag is at least accepted (i.e. no `EINVAL` or any other error is returned):
```
root@kd:~/git/linux# docker run -d busybox top
f24191dc7156767505b674041477f10208847b477dc9ef9f4b1395ed46a62aee
root@kd:~/git/linux# grep nsfs /proc/self/mountinfo /proc/self/mounts
/proc/self/mountinfo:576 26 0:3 net:[4026532644] /run/docker/netns/bf03800bf58e rw shared:271 - nsfs nsfs rw
/proc/self/mounts:nsfs /run/docker/netns/bf03800bf58e nsfs rw 0 0
root@kd:~/git/linux# strace -eumount2 umount -f /run/docker/netns/bf03800bf58e 
umount2("/run/docker/netns/bf03800bf58e", MNT_FORCE) = 0
+++ exited with 0 +++
root@kd:~/git/linux# uname -a
Linux kd 4.14.23+ #46 SMP Fri Mar 2 10:31:35 PST 2018 x86_64 x86_64 x86_64 GNU/Linux
```

From limited reading of linux kernel source code it seems MNT_FORCE is used by any fs which implements `umount_begin` and is ignored otherwise, and it's working that way since Apr 2008 (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=42faad99658ee). The `umount_begin` is defined by 9p, ceph, cifs, fuse, nfs, and lustre (as of linux 4.14).

To sum it up:
1. `MNT_FORCE` is useless (it does nothing for `nsfs`) but harmless (it does not cause an error) for this case. I am slightly in favor of removing it (as it gives a bad example).
2. Yes, both `ENOENT` and `EINVAL` should be ignored.

**Comment by thaJeztah** [comment]
@kolyshkin thanks for doing the research! I updated the PR, and captioned your information in the commit message 👍

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/moby/issues/36547 for the flaky `TestServiceWithPredefinedNetwork`

**Comment by selansen** [comment]
trying to reproduce it. on top this issue now.

**Comment by thaJeztah** [comment]
All of experimental, janky, powerpc and z are failing on `DockerSwarmSuite.TestAPIServiceUpdatePort` and `DockerSwarmSuite.TestSwarmPublishDuplicatePorts` 😞  https://github.com/moby/moby/issues/36501

**Comment by thaJeztah** [comment]
rebased: ping @vdemeester @cpuguy83 @kolyshkin PTAL

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Failure on Windows: https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/20444/console

```
17:54:27 ----------------------------------------------------------------------
17:54:27 FAIL: docker_cli_logs_test.go:41: DockerSuite.TestLogsTimestamps
17:54:27 
17:54:27 docker_cli_logs_test.go:52:
17:54:27     c.Assert(lines, checker.HasLen, testLen+1)
17:54:27 ... obtained []string = []string{"2018-04-17T17:54:26.053926200Z =", ""}
17:54:27 ... n int = 101
17:54:27 
17:54:35 
```

Failure on s390x is odd: https://jenkins.dockerproject.org/job/Docker-PRs-s390x/9397/console

```
17:20:50 ----------------------------------------------------------------------
17:20:50 FAIL: docker_api_containers_test.go:434: DockerSuite.TestContainerAPITop
17:20:50 
17:20:50 docker_api_containers_test.go:452:
17:20:50     c.Assert(top.Processes[0][10], checker.Equals, "/bin/sh -c top")
17:20:50 ... obtained string = "top"
17:20:50 ... expected string = "/bin/sh -c top"
17:20:50 
```

**Comment by thaJeztah** [comment]
Hitting the jackpot with all flaky tests

**Comment by thaJeztah** [comment]
Interesting: all of the Linux ones fail on https://github.com/moby/moby/issues/36501 (`DockerSwarmSuite.TestSwarmPublishDuplicatePorts` and `DockerSwarmSuite.TestServiceUpdatePort`); is this revealing an actual issue?

**Comment by olljanat** [comment]
@thaJeztah this one needs to be rebased

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

File: internal/test/daemon/daemon_unix.go [hunk 2]
```diff
@@ -20,13 +21,29 @@ func cleanupNetworkNamespace(t testingT, execRoot string) {
...
+			continue
+		}
+
+		name := d.Name()
+		path := filepath.Join(netnsPath, name)
+
+		// Unmount the network namespace using MNT_DETACH (lazy unmount).
+		// Ignore EINVAL ("target is not a mount point"), and ENOENT (not found),
+		// which could occur if the network namespace was already gone.
+		if err := unix.Unmount(path, unix.MNT_DETACH); err != nil && err != unix.EINVAL && err != unix.ENOENT {
```
Comment (thaJeztah) [inline internal/test/daemon/daemon_unix.go]: Ah, yes, rebase is needed due to https://github.com/moby/moby/commit/73baee2dcf546b2561bdd9a500b0af08cb62b1be (https://github.com/moby/moby/pull/38127), which largely did the same

I think this PR still has merit, as it removes the use of `filepath.Walk`; let me rebase

**Comment by thaJeztah** [comment]
Test is still failing; need to find out what's causing that;

```
15:10:09 FAIL: docker_cli_swarm_test.go:1534: DockerSwarmSuite.TestSwarmPublishDuplicatePorts
15:10:09 
15:10:09 [dc64601ed788f] waiting for daemon to start
15:10:09 [dc64601ed788f] daemon started
15:10:09 
15:10:09 docker_cli_swarm_test.go:1542:
15:10:09     // make sure task has been deployed.
15:10:09     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, 1)
15:10:09 docker_utils_test.go:441:
15:10:09     c.Assert(v, checker, args...)
15:10:09 ... obtained int = 0
15:10:09 ... expected int = 1
15:10:09 
15:10:09 waited for 30.010791614s (out of 30s)
15:10:09 [dc64601ed788f] exiting daemon
```

Logs attached of that daemon

[docker.log](https://github.com/moby/moby/files/2776544/docker.log)



Test itself looks straightforward at a glance https://github.com/moby/moby/blob/ef91b404efab3f038326ad75ad17d51a8800ea93/integration-cli/docker_cli_swarm_test.go#L1534-L1551

**Comment by thaJeztah** [comment]
The `error creating vxlan interface: file exists` problem was once reported in https://github.com/docker/libnetwork/issues/945 (and https://github.com/docker/libnetwork/issues/562, https://github.com/docker/libnetwork/issues/751 before that)

That case was fixed through https://github.com/docker/libnetwork/pull/1065

**Comment by salzig** [comment]
@thaJeztah just as a note, got this today with docker `18.09.7` on Ubuntu 16.04.

> "network sandbox join failed: subnet sandbox join failed for "10.0.3.0/24": error creating vxlan interface: file exists"

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
This needs to be rebased/retested/merged. @thaJeztah can you please revive it?

**Comment by thaJeztah** [comment]
Rebased to trigger CI again

**Comment by thaJeztah** [comment]
Hmm still failing on this one

```
Stacktrace
=== RUN   TestDockerSwarmSuite/TestSwarmPublishDuplicatePorts
    --- FAIL: TestDockerSwarmSuite/TestSwarmPublishDuplicatePorts (31.39s)
        docker_cli_swarm_test.go:1548: timeout hit after 30s: first check never completed
        poll.go:121: assertion failed: error is not nil: exit status 1

**Comment by thaJeztah** [comment]
Rebased again, to see if things changed since last year

**Comment by guoard** [comment]
@thaJeztah PTAL.

File: internal/testutil/daemon/daemon_linux.go [hunk 3]
```diff
@@ -19,13 +19,32 @@ func cleanupNetworkNamespace(t testing.TB, d *Daemon) {
...
+		path := filepath.Join(netnsPath, dir.Name())
+
+		// Unmount the network namespace using MNT_DETACH (lazy unmount).
+		// Ignore EINVAL ("target is not a mount point"), and ENOENT (not found),
+		// which could occur if the network namespace was already gone.
 		if err := unix.Unmount(path, unix.MNT_DETACH); err != nil && !errors.Is(err, unix.EINVAL) && !errors.Is(err, unix.ENOENT) {
-			t.Logf("[%s] unmount of %s failed: %v", d.id, path, err)
+			t.Logf("[%s] unmount of network namespace %s failed: %v", d.id, path, err)
+		}
+		if err := os.Remove(path); err != nil && !errors.Is(err, os.ErrNotExist) {
```
Comment (kolyshkin) [inline internal/testutil/daemon/daemon_linux.go:44]: Strictly speaking you need to call `syscall.Rmdir` here as we know it's a directory (os.Remove calls Unlink first and Rmdir second). Problem is you need to wrap it into retry-on-EINTR. Something like this:

```go
var err error
for {
	err = unix.Rmdir(path)
	if err != unix.EINTR {
		break
}
if err != nil && err != os.ErrNotExist {
	t.Logf("[%s] unable to remove network namespace %s: %s", d.id, path, err)
}
```

but maybe it's too much for the test.

File: internal/testutil/daemon/daemon_linux.go [hunk 4]
```diff
@@ -19,13 +19,32 @@ func cleanupNetworkNamespace(t testing.TB, d *Daemon) {
...
+			t.Logf("[%s] unable to remove network namespace %s: %s", d.id, path, err)
```
Comment (kolyshkin) [inline internal/testutil/daemon/daemon_linux.go:45]: nit: os.Remove errors already contain the file name, so no need to include `path`.

File: internal/testutil/daemon/daemon_linux.go (hunk 3 above)
Comment (thaJeztah) [reply internal/testutil/daemon/daemon_linux.go:44]: Ah, yeah.

Honestly I just did a rebase of this one, as I was going through old PRs, and this one was still on my curiosity list as to "why is it failing?". It grew more complicated over time, but the original PR was meant to be an easy swap 😂

Comment (kolyshkin) [reply internal/testutil/daemon/daemon_linux.go:44]: It's because there's no `os.RemoveDir` or similar. On both Windows and UNIX syscalls for removing file and directory are different (the only OS supported by Go which doesn't have "RemoveDir" is plan9).
